server/service/src/sensor/berlinger.rs

```diff
@@ -0,0 +1,380 @@
+        };
+        TemperatureBreachRowRepository::new(connection).upsert_one(&new_temperature_breach)?;
+        println!("Added sensor breach {:?} ", new_temperature_breach);
+        Ok(())
+    }
+}
+
+fn sensor_add_breach_config_if_new(
+    connection: &StorageConnection,
+    sensor_row: &SensorRow,
+    temperature_breach_config: &temperature_sensor::TemperatureBreachConfig,
+) -> Result<(), RepositoryError> {
+    let mut config_description = format!(
+        "for {} minutes",
+        temperature_breach_config.duration.num_minutes()
+    );
+    let breach_row_type = get_breach_row_type(&temperature_breach_config.breach_type);
+
+    match temperature_breach_config.breach_type {
+        BreachType::ColdConsecutive => {
+            config_description = format!(
+                "Consecutive {} colder than {}",
+                config_description, temperature_breach_config.minimum_temperature
+            )
+        }
+        BreachType::ColdCumulative => {
+            config_description = format!(
+                "Cumulative {} colder than {}",
+                config_description, temperature_breach_config.minimum_temperature
+            )
+        }
+        BreachType::HotConsecutive => {
+            config_description = format!(
+                "Consecutive {} hotter than {}",
+                config_description, temperature_breach_config.maximum_temperature
+            )
+        }
+        BreachType::HotCumulative => {
+            config_description = format!(
+                "Cumulative {} hotter than {}",
+                config_description, temperature_breach_config.maximum_temperature
+            )
+        }
+    }
+
+    let result =
+        get_matching_sensor_breach_config(connection, &config_description, &breach_row_type)?;
+
+    if let Some(_record) = result.clone().pop() {
+        Ok(())
+    } else {
+        let new_temperature_breach_config = TemperatureBreachConfigRow {
+            id: uuid(),
+            store_id: sensor_row.store_id.clone(),
+            is_active: true,
+            description: config_description.clone(),
+            duration: temperature_breach_config.duration.num_seconds() as i32,
+            r#type: breach_row_type,
+            minimum_temperature: temperature_breach_config.minimum_temperature,
+            maximum_temperature: temperature_breach_config.maximum_temperature,
+        };
+        TemperatureBreachConfigRowRepository::new(connection)
+            .upsert_one(&new_temperature_breach_config)?;
+        println!(
+            "Added sensor breach config {:?} ",
+            new_temperature_breach_config
+        );
+        Ok(())
+    }
+}
+
+fn sensor_add_if_new(
+    connection: &StorageConnection,
+    store_id: &str,
+    temperature_sensor: &temperature_sensor::Sensor,
+) -> Result<(), RepositoryError> {
+    let result = get_matching_sensor_serial(connection, &temperature_sensor.serial)?;
+
+    if let Some(_record) = result.clone().pop() {
```

Comment:
I think this might be better, also for breach config:

https://github.com/openmsupply/open-msupply/compare/1798-Add-support-for-Berlinger-temperature-sensors...1798-Add-support-for-Berlinger-temperature-sensors(Use-is-empty)

Reply:
Thanks for the tip - updated both here: https://github.com/openmsupply/open-msupply/pull/2383/commits/0c811809efa560622c6ed0304ba895f311517d50